### DISCUSSION TITLE
Saving the code pushes to parent widget

### DISCUSSION THREAD

**Discussion by tahamaddam** [fujidaiti/navigator_scope#3] [Q&A]
I am using `navigator_scope` to make the persistent bottom NavigationBar. However, if I am at any level deeper than Home tab, saving code pops back to Home. 

For example, Home Tab > Screen 1. I save the code on Screen 1, it pushes me back to Home Tab. Is this normal behavior? How to prevent it?

**Comment by fujidaiti** [comment]
Could you provide a minimal code to reproduce this issue?

**Comment by tahamaddam** [comment]
Thanks for your reply. 

Here is the code for the class that uses the `navigator_scope`. 

```dart
import 'package:experiments/views/explore/explore_view.dart';
import 'package:experiments/views/library/library_view.dart';
import 'package:experiments/views/settings/settings_view.dart';
import 'package:flutter/material.dart';
import 'package:navigator_scope/navigator_scope.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  int currentTab = 0;

  final tabs = const [
    NavigationDestination(
      icon: Icon(Icons.explore),
      label: 'Explore',
    ),
    NavigationDestination(
      icon: Icon(Icons.library_books),
      label: 'Library',
    ),
    NavigationDestination(
      icon: Icon(Icons.settings),
      label: 'Settings',
    ),
  ];

  final navigatorKeys = [
    GlobalKey<NavigatorState>(debugLabel: 'Explore Tab'),
    GlobalKey<NavigatorState>(debugLabel: 'Library Tab'),
    GlobalKey<NavigatorState>(debugLabel: 'Settings Tab'),
  ];

  NavigatorState get currentNavigator =>
      navigatorKeys[currentTab].currentState!;

  void onTabSelected(int tab) {
    if (tab == currentTab && currentNavigator.canPop()) {
      currentNavigator.popUntil((route) => route.isFirst);
    } else {
      setState(() => currentTab = tab);
    }
  }

  @override
  Widget build(BuildContext context) {
    final body = NavigatorScope(
      currentDestination: currentTab,
      destinationCount: tabs.length,
      destinationBuilder: (context, index) {
        return NestedNavigator(
          navigatorKey: navigatorKeys[index],
          builder: (context) {
            switch (index) {
              case 0:
                return const ExploreView();
              case 1:
                return const LibraryView();
              case 2:
                return const SettingsView();
              default:
                return Container();
            }
          },
        );
      },
    );

    return Scaffold(
      bottomNavigationBar: NavigationBar(
        selectedIndex: currentTab,
        destinations: tabs,
        onDestinationSelected: onTabSelected,
      ),
      body: body,
    );
  }
}

```

I suspect the issue might be related to the GlobalKey.

**Comment by fujidaiti** [comment]
Your code does not appear to have any issues...
Could you please share a simplified version of your code that I can run to reproduce your problem?
Then I can be more helpful.

**Comment by tahamaddam** [comment]
It is hard to share a simplified version, I thought the problem might be related to the way NavigationBar is handled as the app was not causing the issue before this NavigationBar implementation. I am using various Stream and StreamBuilder, mostly from Firebase instances. I am not sure If that is causing the Issue.

 If it is of any help, this is the code for MyApp (one layer top ofHomeScreen that I shared above)

```dart
class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Experiments',
        theme: FlexThemeData.light(
          scheme: FlexScheme.blue,
          useMaterial3: true,
        ),
        darkTheme:
            FlexThemeData.dark(scheme: FlexScheme.blue, useMaterial3: true),
        themeMode: ThemeMode.system,
        home: StreamBuilder<User?>(
          stream: FirebaseAuth.instance.authStateChanges(),
          builder: (BuildContext context, AsyncSnapshot<User?> snapshot) {
            if (snapshot.connectionState == ConnectionState.active) {
              final User? user = snapshot.data;
              if (user == null) {
                return const LoginScreen();
              } else {
                if (!user.emailVerified) {
                  return const VerifyEmailScreen();
                } else {
                  return const HomeScreen();
                }
              }
            }
            return const CircularProgressIndicator();
          },
        ),
        routes: {
          '/login': (context) => const LoginScreen(),
          '/home_screen': (context) => const HomeScreen(),
          '/verify-email': (context) => const VerifyEmailScreen(),
        });
  }
}
```

**Reply by fujidaiti** [reply]
What is the result of `snapshot.connectionState == ConnectionState.active` after saving file?
I don't know much about your project so this is my guess:

1. edit and save files
2. hot reloading runs
3. the whole widget tree rebuilds
4. `snapshot.connectionState == ConnectionState.active` being false at the first build (because of some async process?)
5. `CircularProgressIndicator` is displayed and `HomeScreen` is removed from the widget tree
6. `snapshot.connectionState == ConnectionState.active` being true right after the first frame (so the indicator is only visible for the very short time)
7. `HomeScreen` is shown but the state is refreshed

I hope this helps you.

**Comment by tahamaddam** [comment]
Thanks for your time for looking into this, I debugPrint the states and the problem seems to be the one you elaborated above.  So for the time of testing, I removed the conditions for `home` parameter and set it directly to HomeScreen() which solved the issue of view popping back to main.  Thanks again. 

If you have any better solution for managing the conditions for `home` during the startup that avoids this pop-to-home issues, please do let me know. Thanks.

**Reply by fujidaiti** [reply]
>solution for managing the conditions for home during the startup that avoids this pop-to-home issue

I've had the same problem before. Rather than swapping them, I would recommend stacking the loading page on top of the home page, or showing the loading indicator within the home page itself.

Hope this works, thanks.

**Comment by fujidaiti** [comment]
Close as not related to the package.
